cmd/utils.go

```diff
@@ -108,6 +103,26 @@ func (handler *SignalHandler) RegisterWithoutExiting() {
 	for _, callback := range handler.callbacks {
 		callback()
 	}
+	os.Exit(0)
+}
+
+// RegisterWithContext is a no-exit version of Register function with context usage
+func (handler *SignalHandler) RegisterWithContext(globalContext context.Context) {
+	signal.Notify(handler.ch, handler.signals...)
+	for {
+		select {
+		case <-handler.ch:
+			for _, listener := range handler.listeners {
+				listener.Close()
+			}
+			for _, callback := range handler.callbacks {
+				callback()
+			}
+		case <-globalContext.Done():
+			// got signal for shutdown, so just return
+			return
```

Comment:
Do I understand it right that it is not necessary to close any listeners in this case because they will be closed by AcraServer when it sees that the context is Done()?

Reply:
Not necessary, because some other component initialized exit from the program and here we just notify the background goroutine that it should stop the execution, to avoid any background actions.

For example, when acra-server receives SIGTERM, sigterm handler stops all components and should notify all goroutines that they should stop any processing to finalize log. It will raise SIGUSR1 handler too, but it should not do anything except finish goroutine.